content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -419,6 +411,701 @@ If :ada:`X (Index)` occurs inside :ada:`Process_Array`, there is no need
+          WHR.Register (Int_Stor (I).all.Show'Access);
+       end loop;
+
+       --  Now, use Process_All to call the handlers
+       --  (in this case, the Show procedure for
+       --  each protected object from Int_Stor).
+
+       WHR.Process_All;
+
+    end Show_Access_To_Protected_Subprograms;
+
+The work handler registry (:ada:`WHR`) has a maximum capacity of five
+procedures, whereas the :ada:`Int_Stor` array has a capacity of three elements.
+By calling :ada:`WHR.Register` and passing :ada:`Int_Stor (I).all.Show'Access`,
+we register the :ada:`Show` procedure of each protected object from
+:ada:`Int_Stor`.
+
+.. admonition:: Important
+
+    Note that the components of the :ada:`Int_Stor` array are of
+    :ada:`Integer_Storage_Access` type, which is declared as an access to
+    :ada:`Integer_Storage` objects. Therefore, we have to dereference the
+    object (by writing :ada:`Int_Stor (I).all`) before getting access to the
+    :ada:`Show` procedure (by writing :ada:`.Show'Access`).
+
+    We have to use an access type here because we cannot pass the access (to
+    the :ada:`Show` procedure) of a local object in the call to the
+    :ada:`Register` procedure. Therefore, the protected objects (of
+    :ada:`Integer_Storage` type) cannot be local.
+
+    This issue becomes evident if we replace the declaration of
+    :ada:`Int_Stor` with a local array (and then adapt the remaining code). If
+    we do this, we get a compilation error in the call to :ada:`Register`:
+
+    .. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Protected_Access_Init_Function
+        :class: ada-expect-compile-error
+
+        with Work_Registry;          use Work_Registry;
+        with Integer_Storage_System; use Integer_Storage_System;
+
+        procedure Show_Access_To_Protected_Subprograms is
+
+           WHR      : Work_Handler_Registry (5);
+
+           Int_Stor : array (1 .. 3) of Integer_Storage;
+
+        begin
+           --  Allocate and initialize integer storage
+           --
+           --  (For the initialization, we're just assigning
+           --   the index here, but we could really have used
+           --   any integer value.)
+
+           for I in Int_Stor'Range loop
+              --  Int_Stor (I) := new Integer_Storage;
+              Int_Stor (I).Set (I);
+           end loop;
+
+           --  Register handlers
+
+           for I in Int_Stor'Range loop
+              WHR.Register (Int_Stor (I).Show'Access);
+              --            ^ ERROR!
+           end loop;
+
+           --  Now, call the handlers
+           --  (i.e. the Show procedure of each
+           --   protected object).
+
+           WHR.Process_All;
+
+        end Show_Access_To_Protected_Subprograms;
+
+    As we've just discussed, this error is due to the fact that
+    :ada:`Int_Stor` is now a "local" protected object, and the accessibility
+    rules don't allow mixing it with non-local accesses.
```

Comment:
I suggest adding " in order to prevent the possibility of dangling references." to the end of the last sentence above.